### DISCUSSION TITLE
How to "re-configure cert-manager" to use AAD Workload Identity in deployKF

### DISCUSSION THREAD

**Discussion by edi-bice** [deployKF/deployKF#58] [Questions & Answers] [answered]
The instructions here assume Helm - not sure that is the case with deployKF

https://cert-manager.io/docs/configuration/acme/dns01/azuredns/#reconfigure-cert-manager

extraManifests and cert_manager.controller ... annotations seem promising

**Comment by thesuperzapper** [comment, marked as answer]
@edi-bice thank you for reminding me! 

To use Azure Workload Identity for cert-manager authentication, we need to add new values that let you set, ~ServiceAccount LABELS (not annotations)~ and Pod LABELS.

EDIT: after checking the Azure docs, I think there is a typo in the [cert-manager AKS docs](https://cert-manager.io/docs/configuration/acme/dns01/azuredns/#reconfigure-cert-manager), because [the required labels/annotations](https://azure.github.io/azure-workload-identity/docs/topics/service-account-labels-and-annotations.html) are:

- Pod - LABEL: `azure.workload.identity/use: "true"`
- ServiceAccount - ANNOTATION: `azure.workload.identity/client-id: "...."`

Currently, we only provide [`deploykf_dependencies.cert_manager.controller.serviceAccount.annotations`](https://github.com/deployKF/deployKF/blob/v0.1.3/generator/default_values.yaml#L156), which sets ServiceAccount annotations. It should be relatively straightforward for us to add new values in the next release:

- ~`deploykf_dependencies.cert_manager.controller.serviceAccount.labels`~
- `deploykf_dependencies.cert_manager.controller.podLabels`

---

However, you can use the internal helm chart values before then with the `deploykf_dependencies.cert_manager.valuesOverrides` value (we use the upstream cert-manager helm chart internally).

For example:

```yaml
deploykf_dependencies:
  cert_manager:
    valuesOverrides: |
      cert-manager:
        podLabels:
          azure.workload.identity/use: "true"

    controller:
      serviceAccount:
        annotations:
          azure.workload.identity/client-id: "00000000-0000-0000-0000-000000000000"

          ## only needed if want a non-default AZURE_TENANT_ID
          #azure.workload.identity/tenant-id: "00000000-0000-0000-0000-000000000000"
```

**Reply by thesuperzapper** [reply]
I have made an issue to track this for the next release: https://github.com/deployKF/deployKF/issues/59

**Reply by edi-bice** [reply]
I added the stanza above in the cert-manager section of the sample-values-overrides.yaml and did ./scripts/sync_argocd_apps.sh multiple times, followed by multiple rollout restarts of cert-manager and cert-manager-webhook deployments in cert-manager namespace.

Regardless of what I tried, `kubectl describe pod -n cert-manager -l app.kubernetes.io/component=controller` did not show any of the expected environment variables shown here https://cert-manager.io/docs/configuration/acme/dns01/azuredns/

After reading about the mutating webhook role and its prerequisites, I did export the mentioned environment variable via `export AZURE_TENANT_ID="$(az account show -s <AzureSubscriptionID> --query tenantId -otsv)"` and resynced though cert-manager-webhook
continues to be OutOfSync (according to ArgoCD dashboard) and the pod describe above shows none of the env variables.

The webhookd documentation https://azure.github.io/azure-workload-identity/docs/installation/mutating-admission-webhook.html shows the AZURE_TENANT_ID environment variable being used in the helm chart install but grepping for it through deployKF codebase returns nothing

**Reply by thesuperzapper** [reply]
@edi-bice I noticed that the cert-manager docs are actually wrong, and you only need to set a Pod Label for `azure.workload.identity/use`, and you can specify the `azure.workload.identity/client-id` with an ANNOTATION on the ServiceAccount (I have updated the initial response with example values above).

Can you also check why ArgoCD is showing the `dkf-dep--cert-manager` app as out-of-sync, by looking at the app diff in the UI?

**Reply by edi-bice** [reply]
Applications/dkf-dep--cert-manager shows App Health as Healthy, Sync Status as OutOfSync and LastSync as Sync OK. Clicking on Applications/dkf-dep--cert-manage App Diff brings up ValidatingWebhookConfiguration/cert-manager-webhook compact diff, with 

```
- key: control-plane
   operator: NotIn
   values:
      - 'true'
- key: kubernetes.azure.com/managedby
   operator: NotIn
   values:
      - aks
```

cert-manager pod logs show "error retrieving resource lock kube-system/cert-manager-controller net/http: TLS handshake timeout"  

Also of note is that dkf-dep--istio shows the same OutOfSync behavior as dkf-dep--cert-manager in MutatingWebhookConfiguration/istio-sidecar-injector

**Reply by thesuperzapper** [reply]
@edi-bice that out-of-sync behavior is caused by a strange thing that AKS does called the [`Admission Enforcer`](https://learn.microsoft.com/en-us/azure/aks/faq#can-admission-controller-webhooks-impact-kube-system-and-internal-aks-namespaces). It's been discussed upstream [in the cert-manager repo](https://github.com/cert-manager/cert-manager/issues/4114#issuecomment-1145060432), and you need to label OR annotate your webhooks with `admissions.enforcer/disabled: "true"`.

I have been digging into it more, and I really think that Microsoft needs to remove this feature from AKS, because it's just a bad idea to randomly mutate resources submitted to the cluster, because it inevitably gets into fights with other controllers or GitOps tools.

I have raised an issue with AKS about giving a global off switch for this feature, which you can __promote with your Microsoft representative__: https://github.com/Azure/AKS/issues/4002

Microsoft did at least say they would remove one part of the admission enforcer a few weeks ago in https://github.com/Azure/AKS/issues/177, but I think we need an option to fully disable it.

---

Annoyingly, because of this, to get deployKF working on AKS there are several things you will have to do (until I can make some of these changes by default).

The following apps need the annotation for actual functional reasons (not just because of the argocd diff):

- __kyverno__ (https://github.com/kyverno/kyverno/issues/6539)
    - annotation can be added via existing helm values
    - this one will cause a memory leak if not annotated
- __kubeflow-pipelines__ (https://github.com/kubeflow/pipelines/issues/5244)
    - can NOT currently be added using existing values
    - this one could also be made into a "diff only" issue if I remove the [`control-plane` label from the `kubeflow` namespace](https://github.com/deployKF/deployKF/blob/v0.1.3/generator/templates/argocd/namespaces/kubeflow-tools/kubeflow.yaml#L6)

The remaining ones are "cosmetic only", and can either be solved by the annotation or telling argocd to ignore the diff:

- __cert-manager / trust-manager__ 
    - annotation can be added via existing helm values
- __istio__
     - annotation can be added via existing helm values
- __kubeflow katib__ 
    - can NOT currently be added using existing values
- __kubeflow poddefaults__ 
    - can NOT currently be added using existing values

I am not sure if I should just include the annotations by default in deployKF, as they should not impact other users, or if I should just expose them all as values, and show them in a "deployKF on AKS" guide, similar to the one I am writing for EKS.

---

For reference, at least for cert-manager, you could sort of fix it by adding additional value overrides like this:

```yaml
deploykf_dependencies:
  cert_manager:
    valuesOverrides: |
      cert-manager:
      
        ## add this
        webhook:
          validatingWebhookConfigurationAnnotations:
            admissions.enforcer/disabled: "true"
```

**Reply by thesuperzapper** [reply]
I have raised a separate deployKF issue to track issues with AKS admission enforcer:
 - https://github.com/deployKF/deployKF/issues/61

As this is separate from adding values for Azure Workload Identity.

**Reply by edi-bice** [reply]
I added the webhook stanza above as well as removed the control-plane label from kubeflow namespace, rollout restarted all deployments in all namespaces and did another ./scripts/sync_argocd_apps.sh yet both dkf-dep--cert-manager and dkf-dep--istio remain OutOfSync. Will add the "admissions.enforcer/disabled: true" to other webhooks and see. BTW, earlier I used this https://azure.github.io/kubeflow-aks/main/ to deploy on AKS and was able to get things up and running with just one patch applying the same annotation to the istio-sidecar-injector MutatingWebhookConfiguration. I will see if the AKS-Construction tool provides options to turn off the admission enforcer logic though it seems from your research and the myriad of tickets that it is not yet exposed/supported

**Reply by edi-bice** [reply]
I was able to deploy successfully after changing to generator mode, basically generating the yaml, checking them in to git and specifying repo and branch to reconcile

**Reply by thesuperzapper** [reply]
@edi-bice Just following up to say [deployKF v0.1.4](https://github.com/deployKF/deployKF/releases/tag/v0.1.4) now works properly with Azure AKS:

- __Admission Enforcer:__
    - https://github.com/deployKF/deployKF/issues/61#issuecomment-1949658332
- __Cert Manager Labels:__
    - https://github.com/deployKF/deployKF/pull/88

To use deployKF with AKS, you need to set the `kubernetes.azure.admissionsEnforcerFix` value to `true`:

```yaml
kubernetes:
  azure:
    admissionsEnforcerFix: true
```
